compose/koin-androidx-compose/src/main/java/org/koin/androidx/compose/KoinApplication.kt

```diff
@@ -0,0 +1,64 @@
+/*
+ * Copyright 2017-2023 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+@file:OptIn(KoinInternalApi::class)
+
+package org.koin.androidx.compose
+
+import android.app.Application
+import android.content.ComponentCallbacks
+import android.content.Context
+import android.content.ContextWrapper
+import androidx.compose.runtime.Composable
+import androidx.compose.runtime.CompositionLocalProvider
+import androidx.compose.runtime.remember
+import androidx.compose.ui.platform.LocalContext
+import org.koin.android.ext.android.getKoin
+import org.koin.compose.LocalKoinApplication
+import org.koin.compose.LocalKoinScope
+import org.koin.core.annotation.KoinInternalApi
+import org.koin.core.component.KoinComponent
+
+/**
+ * Provide active Koin application from Android [Context] to Compose
+ *
+ * @param content - following compose function
+ *
+ * @author Jan-Jelle Kester
+ */
+@Composable
+fun KoinApplication(content: @Composable () -> Unit) {
```

Comment:
The `koin-compose` module has already a `KoinApplication` function. Do we need a "special" one here?

Reply:
For this specific use-case, yes.

The `KoinApplication` function that already exists _creates_ a new application and starts it. For this one, the assumption is that externally (likely in the Android `Application` class) an application was already created. This is an Android specific implementation because it uses the Android `Context` to find the relevant Koin application/scope. This is all not implemented for multiplatform Compose (as it works differently), so this code cannot live in `koin-compose`.

In short, this is needed to work around the issue that the composition local defaults are never changing between test runs, ensuring that we have a way of setting them to the correct value.

Personally I don't like this workaround, since we are introducing production code purely for test purposes, but with the way composition locals in Compose are built and the way Koin uses them (for composition-based scopes) I could not think of a good alternative.